janusgraph-hadoop-parent/janusgraph-hadoop-core/src/main/java/org/janusgraph/hadoop/formats/cql/CqlBinaryRecordReader.java

```diff
@@ -0,0 +1,122 @@
+    private KV currentKV;
+    private KV incompleteKV;
+
+    private final CqlRecordReader reader;
+
+    public CqlBinaryRecordReader(CqlRecordReader reader) {
+        this.reader = reader;
+    }
+
+    @Override
+    public void initialize(final InputSplit inputSplit, final TaskAttemptContext taskAttemptContext) throws IOException {
+        reader.initialize(inputSplit, taskAttemptContext);
+    }
+
+    @Override
+    public boolean nextKeyValue() throws IOException {
+        return null != (currentKV = completeNextKV());
+    }
+
+    private KV completeNextKV() throws IOException {
+        KV completedKV = null;
+        boolean hasNext;
+        do {
+            hasNext = reader.nextKeyValue();
+
+            if (!hasNext) {
+                completedKV = incompleteKV;
+                incompleteKV = null;
+            } else {
+                Row row = reader.getCurrentValue();
+                StaticArrayBuffer key = StaticArrayBuffer.of(row.getBytesUnsafe(CQLKeyColumnValueStore.KEY_COLUMN_NAME));
+                StaticBuffer column1 = StaticArrayBuffer.of(row.getBytesUnsafe(CQLKeyColumnValueStore.COLUMN_COLUMN_NAME));
+                StaticBuffer value = StaticArrayBuffer.of(row.getBytesUnsafe(CQLKeyColumnValueStore.VALUE_COLUMN_NAME));
+                Entry entry = StaticArrayEntry.of(column1, value);
+
+                if (null == incompleteKV) {
+                    // Initialization; this should happen just once in an instance's lifetime
+                    incompleteKV = new KV(key);
+                } else if (!incompleteKV.key.equals(key)) {
+                    // The underlying Cassandra reader has just changed to a key we haven't seen yet
+                    // This implies that there will be no more entries for the prior key
+                    completedKV = incompleteKV;
+                    incompleteKV = new KV(key);
+                }
+
+                incompleteKV.addEntry(entry);
+            }
+            /* Loop ends when either
+             * A) the cassandra reader ran out of data
+             * or
+             * B) the cassandra reader switched keys, thereby completing a KV */
+        } while (hasNext && null == completedKV);
+
+        return completedKV;
+    }
+
+    @Override
+    public StaticBuffer getCurrentKey() {
+        return currentKV.key;
+    }
+
+    @Override
+    public Iterable<Entry> getCurrentValue() {
+        return currentKV.entries;
+    }
+
+    @Override
+    public void close() {
+        reader.close();
+    }
+
+    @Override
+    public float getProgress() {
+        return reader.getProgress();
+    }
+
+    private static class KV {
+        private final StaticArrayBuffer key;
+        private ArrayList<Entry> entries;
+
+        public KV(StaticArrayBuffer key) {
+            this.key = key;
+        }
+
+        public void addEntry(Entry toAdd) {
+            if (null == entries)
```

Comment:
Is there cases where you create `KV` but don't call `addEntry` for it?

Also. I would prefer braces `{}` where possible.

Reply:
If we always call `addEntry` for new KV then move `new ArrayList<>();` to class variable initialization or constructor.

Reply:
I think we copied the `KV` class from another input format, but your suggestion makes sense.